### ISSUE TITLE
Product of unbounded intervals with intervals containing 0

### ISSUE THREAD

**Issue by clslgrnc** [mpmath/mpmath#591]
Currently:
```pycon
>>> from mpmath import iv
>>> iv.mpf(['0', '0']) * iv.mpf(['1', 'inf'])
mpi('-inf', '+inf')
>>> iv.mpf(['0', '1']) * iv.mpf(['1', 'inf'])
mpi('0.0', '+inf')
>>> iv.mpf(['-1', '0']) * iv.mpf(['1', 'inf'])
mpi('-inf', '0.0')
```

There is either a bug or a missed opportunity for improved precision. Indeed for any three sets A, B, and C and any function f, we have:

- { f(x, y) ∣ x ∈ A ⋂ B and y ∈ C } ⊆ { f(x, y) ∣ x ∈ A and y ∈ C }
- and { f(x, y) ∣ x ∈ A ⋂ B and y ∈ C } ⊆ { f(x, y) ∣ x ∈ B and y ∈ C }
- thus { f(x, y) ∣ x ∈ A ⋂ B and y ∈ C } ⊆ { f(x, y) ∣ x ∈ A and y ∈ C } ⋂ { f(x, y) ∣ x ∈ B and y ∈ C }

Taking A = iv.mpf(['0', '1']), B = iv.mpf(['-1', '0']), C = iv.mpf(['1', 'inf']), and f being the multiplication we get:
iv.mpf(['0', '0']) * iv.mpf(['1', 'inf']) ⊆ mpi('0.0', '+inf') ⋂ mpi('-inf', '0.0') = mpi('0.0', '0.0')

According to @skirpichev `iv.mpf(['0', '0']) * iv.mpf(['1', 'inf'])` can not be `mpi('0.0', '0.0')` because mpmath intervals are defined over extended reals with undefined 0*∞ (see #334).
Then this should be explicit in the documentation and the results of the product of any interval containing 0 with any interval containing ∞ should have a consistent value.

**Comment by skirpichev** [comment]
Just for the record, that's Mathematica output:
```
In[4]:= Interval[{0, 0}] Interval[{1, Infinity}]

Infinity::indet: Indeterminate expression 0 Infinity encountered.

Out[4]= Interval[{0, Indeterminate}]
```
I doubt we have options here, given the mpmath uses [usual definition](https://en.wikipedia.org/wiki/Interval_arithmetic#Interval_operators) for multiplication.

**Comment by clslgrnc** [comment]
> Just for the record, that's Mathematica output:
> 
> ```
> In[4]:= Interval[{0, 0}] Interval[{1, Infinity}]
> 
> Infinity::indet: Indeterminate expression 0 Infinity encountered.
> 
> Out[4]= Interval[{0, Indeterminate}]
> ```
> 

Indeed. I do not have access to Mathematica but I guess WolframAlpha Open Code uses the same engine.
Interestingly, it returns the same results as mpmath for `[0, 1]*[1, inf]` and `[-1, 0]*[1, inf]`.

Returning `[-inf, inf]` or a warning and an empty set are valid choices. I am just arguing that if the results for `[0, 1]*[1, inf]` and `[-1, 0]*[1, inf]` are correct, one can do better.

For any two sets A and B such that A is a subset of B, and any function f:

- if f is defined over B then f is defined over A
- and { f(x) : x in A } is a subset of { f(x) : x in B }

Do you agree so far?

Taking:

- A the cartesian product of [0, 0] and [1, inf]
- B the cartesian product of [0, 1] and [1, inf]
- f the multiplication operator

Then we have that `[0, 0]*[1, inf]` is included in `[0, 1]*[1, inf]`.
And  similarly `[0, 0]*[1, inf]` is included in `[-1, 0]*[1, inf]`

One could argue that f is not defined over B, but it seems to be since mpmath and Mathematica both return a valid interval without emitting any warning containing by definition all x*y for x in [0, 1] and y in [1, inf].

Do you agree so far?

As a result:

- either `[0, 0]*[1, inf] = [0, 0]`
- or mpmath and Mathematica give incorrect results for `[0, 1]*[1, inf]` and `[-1, 0]*[1, inf]`
- (or maybe I made a mistake and I'll be glad if anyone points it out to me)

> I doubt we have options here,

Actually there are [quite a few](https://www.boost.org/doc/libs/1_76_0/libs/numeric/interval/doc/checking.htm).
And one of them is to just drop invalid values, silently or not. That would mean in my previous argument that f is indeed not defined over B:

- Matematica should be more consistent in its warnings
- mpmath could return an empty set for `[0, 0]*[1, inf]`

> given the mpmath uses [usual definition](https://en.wikipedia.org/wiki/Interval_arithmetic#Interval_operators) for multiplication.

You'll notice that for the simplification to be valid the binary operator must be defined over the whole intervals.
Yet in mpmath's intervals, it is my understanding that 0*inf is undefined. 

Anyway, interval arithmetic is not the focus of mpmath and the documentation emphasizes that it is still experimental and should be used with caution. If there is not enough time or energy to devote to this issue (I certainly do not have much more), this would be perfectly understandable.

**Comment by skirpichev** [comment]
> I do not have access to Mathematica but I guess WolframAlpha Open Code uses the same engine.

You also can reproduce this with free-access Wolfram Language Engine.

> Returning [-inf, inf] or a warning and an empty set are valid choices.

Warning - definitely not an option.  Maybe raise an exception.

But I think that current approach is fine.

> Yet in mpmath's intervals, it is my understanding that 0*inf is undefined.

With usual floats, like mpf, - yes, we have no option but
```pycon
>>> 0 * inf
mpf('nan')
```
But for iv's we can more refine nan's meaning.  Say, not use one for something like "out of domain range for the given function".

> interval arithmetic is not the focus of mpmath and the documentation emphasizes that it is still experimental and should be used with caution.

I would appreciate a *concrete* proposal for improving documentation.  Meanwhile, I'll close this as "not planned".
